src/models/m_wikiconcept.erl

```diff
@@ -290,6 +299,92 @@ filter_lang(#trans{ tr = Tr }, Context) ->
     #trans{ tr = Tr2 }.
 
 
+%% @doc Handle the merge of two keywords. Merge all connected concepts.
+-spec rsc_merge(WinnerId, LoserId, Context) -> ok when
+    WinnerId :: m_rsc:resource_id(),
+    LoserId :: m_rsc:resource_id(),
+    Context :: z:context().
+rsc_merge(WinnerId, LoserId, Context) ->
+    case z_db:q("
+        update wikiconcept
+        set keyword_id = $1
+        where keyword_id = $2",
+        [ WinnerId, LoserId ],
+        Context)
+    of
+        0 ->
+            ok;
+        Count when is_integer(Count) ->
+            publish_update(WinnerId, Context)
+    end.
+
+%% @doc Find the keyword best matching to the given concept.
+%% First tries to find a keyword in the ascendants of a concept,
+%% then tries to find it in the descendents.
+-spec find_keyword(ConceptID, Context) -> {ok, KeywordRscId} | {error, Reason} when
+    ConceptID :: binary() | string(),
+    Context :: z:context(),
+    KeywordRscId :: m_rsc:resource_id(),
+    Reason :: term().
+find_keyword(ConceptID, Context) ->
+    case get_concept(ConceptID, Context) of
+        {ok, #{
+            <<"keyword_id">> := RscId
+        }} when is_integer(RscId) ->
+            {ok, RscId};
+        {ok, #{
+            <<"wikidata_id">> := WikidataID,
+            <<"wikidata_ancestor_ids">> := Ancestors
+        }} ->
+            case find_nearest_ancestor(Ancestors, Context) of
+                {ok, RscId} ->
+                    {ok, RscId};
+                {error, enoent} ->
+                    find_nearest_descendent(WikidataID, Context)
+            end;
+        {error, _} = Error ->
+            Error
+    end.
+
+find_nearest_ancestor(undefined, _Context) ->
+    {error, enoent};
+find_nearest_ancestor([], _Context) ->
+    {error, enoent};
+find_nearest_ancestor(Ancestors, Context) ->
+    case z_db:q1("
+        select keyword_id
+        from wikiconcept
+        where keyword_id is not null
+          and wikidata_id in (select(unnest($1::character varying[])))
+        order by level desc
+        limit 1",
+        [ Ancestors ],
+        Context)
+    of
+        undefined ->
+            {error, enoent};
+        RscId ->
+            {ok, RscId}
+    end.
+
+find_nearest_descendent(WikidataID, Context) ->
+    case z_db:q1("
+        select keyword_id
+        from wikiconcept
+        where keyword_id is not null
+          and wikidata_ancestor_ids && $1
+        order by level asc
```

Comment:
I'm confused as to why this isn't also:
```suggestion
        order by level desc
```
IIUC when finding the nearest ancestor we get the one with the highest `level` because among those that's the most general (or less specific) one.
IOW we get the highest and furthest ancestor above (level-wise) the concept.

Here however we get, among the descendents, the one with the lowest `level`, so the most specific (or less general).
IOW we get the lowest and furthest descendent below (level-wise) the concept.

In both cases we get the furthest related concept, but perhaps in both cases we should get the highest one, it being the most general that is in the same branch of the tree as the input concept.

Reply:
I think we are looking _up_ we should look for the most specific (ie. highest level) concept.

When we are looking _down_ then we should look to the most generic one, being probably the one _closest_ to the concept we are searching from.

Reply:
oh, I misunderstood, I thought that the most generic concepts had the highest levels.

Thanks for clarifying that it's the other way around.
In this case I think that you're right and this is the best way to go 👍